src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -218,6 +220,50 @@ public UnfilteredRowIterator executePartitionReadCommand(SinglePartitionReadComm
         return command.queryMemtableAndDisk(cfs, executionController);
     }
 
+    public RangeIterator buildIterator()
+    {
+        var filterOperation = filterOperation();
+        var orderings = filterOperation.expressions()
+                                       .stream().filter(e -> e.operator() == Operator.ANN).collect(Collectors.toList());
+        assert orderings.size() <= 1;
+        if (filterOperation.expressions().size() == 1 && filterOperation.children().isEmpty() && orderings.size() == 1)
+            // If we only have one expression, we just use the ANN index to order and limit.
+            return getTopKRows(orderings.get(0));
+
+        // We already decided we need to do first sort then filter, so no need to open the index iterator:
+        if (queryContext.filterSortOrder() == QueryContext.FilterSortOrder.SORT_THEN_FILTER)
+        {
+            assert queryContext.postFilterSelectivityEstimate() != null;
+            return getTopKRows(orderings.get(0));
+        }
+
+        var nonOrderingExpressions = filterOperation.expressions().stream()
+                                                    .filter(e -> e.operator() != Operator.ANN)
+                                                    .collect(Collectors.toList());
+        var iter = Operation.Node.buildTree(nonOrderingExpressions, filterOperation.children(), filterOperation.isDisjunction()).analyzeTree(this).rangeIterator(this);
+        queryContext.setPostFilterSelectivityEstimate(estimateSelectivity(iter));
+
+        if (orderings.isEmpty())
+            return iter;
+
+        // TODO: compute proper threshold
+        if (queryContext.filterSortOrder() == null && iter.getMaxKeys() >= 0)
```

Comment:
Nit: if the max keys is 0, we have disjoint predicates, so we can exit early with an empty iterator.